### DISCUSSION TITLE
Module Federation `@` syntax support

### DISCUSSION THREAD

**Discussion by zamotany** [callstack/repack#199] [Ideas]
Having a `app@remoteUrl` syntax support when declaring remotes, for example:
```js
new webpack.container.ModuleFederationPlugin({
  remotes: {
    app1: Repack.Federated.createRemote('app1'),
    app2: Repack.Federated.createRemote('app2@https://somewhere-on-internet.com/app2.container.bundle'),
  },
})
```
essentially boils down to passing the `remoteUrl` into the `ScriptManager`.

By introducing concept of multiple resolvers to `ScriptManager` this can be easily achieved:
```js
export function createRemote(remote: string) {
    const [remoteName, url] = remote.split('@');

    return dedent`promise new Promise((resolve, reject) => {
      var scriptManager = __webpack_require__.repack.shared.scriptManager;

      if (url) {
        scriptManager.addResolver(function(scriptId) {
          if (scriptId === "${remoteName}") {
            return { url: "${url}" };
          }
        });
      }

      scriptManager
        .loadScript('${remoteName}', undefined, __webpack_require__)
        .then(function() {
          resolve({
            get: (request) => {
              return self.${remoteName}.get(request);
            },
            init: (arg) => {
              try {
                return self.${remoteName}.init(arg);
              } catch (e) {
                console.log('[Repack/Federated] Remote container ${remoteName} already initialized.');
              }
            }
          });
        })
        .catch(function(reason) {
          reject(reason);
        });
    })`;
  }
```

This will allow not only to support `@` syntax in remotes but also add an ability for containers to dynamically provide resolves inside the code:
```js
// In a container

ScriptManager.shared.addResolver(async (scriptId) => {
  if (scriptId === 'app1') {
    const url = await getUrlDynamically();
    return { url };
  }
}):
```

cc: @ScriptedAlchemy @tom-sherman

**Comment by ScriptedAlchemy** [comment]
Yeah this is what id had in mind and tried with v2, but ran into issues of ChunkManager never resolving a promise even though the script did inject into RN. 

Support of the @ syntax would enable pretty much universal use of MFP across any environment with more or less the same syntax for all - just different implementations of the "last mile" interconnections. 

We have this for node.js, next.js which gives us server side MFP capabilities. RN will be the last env to achieve the trifecta. 

What eager issues do you run into? 
When using promise new promise, does the remote get loaded and its just a sharing issue? Is it the host failing on eager error, or is it the now loaded remote that says there is no shared modules injected into it? 


Meta suggestion - what i usually do is create a new plugin wrapper. like RNModuleFederationPlugin, where my remotes are just the normal "browser" synax options. like a straight string - then in the plugin i parse options and convert it to promise new promise. That way the user only needs to pass one syntax all the time and the plugin takes the standard options and can convert it on demand. 

so regardless of my envs, user only needs to read webpack docs and pass a remote in one of the valid formats, i then wrap the real MF plugin in the converted options. 

See: Configuration Options

https://app.privjs.com/buy/packageDetail?pkg=@module-federation/nextjs-ssr

Next is powered by promise new promise, but users only ever pass known ways of working with standard api.

That said, i am working on this: https://github.com/module-federation/enhanced
Which might be a nicer syntax to take advantage of in general for advanced use cases where we need more plugability or middlewares

**Reply by zamotany** [reply]
The eager issue happens only in the host application with `remotes`, for example:
```
// host
new webpack.container.ModuleFederationPlugin({
  name: 'host',
  shared: {
    react: {
      singleton: true,
      eager: true,
    },
    'react-native': {
      singleton: true,
      eager: true,
    },
  },
  remotes: {
    app: `promise new Promise(...)` // or Federated.createRemote, the implementation is a detail here
  }
})
```

Then:
1. Webpack evaluates entrypoints, one of them being `InitializeCore` from `react-native`
2. Inside `InitializeCore` there's an `__webpack_require__` for `webpack/sharing/consume/default/react-native/react-native`
3. This import causes `default` sharing scope to be initialized which has call to `initExternal('app')`
4. Scope initialization return a Promise, because `app` is `promise new Promise`
5. Then this condition is hit and throws:
   ```
   if(typeof factory !== "function") throw new Error("Shared module is not available for eager consumption: " + id);
   ```

**Reply by tom-sherman** [reply]
Have you tried an additional bootstrap step?

index.js:

```js
import Bootstrap from "./bootstrap";

AppRegistry.registerComponent(appName, () => Bootstrap);
```

bootstrap.js:

```js
import * as React from "react";
import { Text } from "react-native";

const App = React.lazy(() => import("./App"));

export default function Bootstrap() {
  return (
    <React.Suspense
      fallback={
        <Text style={{ textAlign: "center", marginTop: 40 }}>
          Loading the app...
        </Text>
      }
    >
      <App />
    </React.Suspense>
  );
}
```

**Reply by ScriptedAlchemy** [reply]
Okay so it looks like it might be that share scope isn't passed into a remote. But the host still works?

If that's the case. I trick I use in next, where I can't move things around in the module graph. Is object.assign(__webpack_share_scopes__.default,{moreStuff})

**Reply by zamotany** [reply]
@tom-sherman
Yup, I've doing it with bootstrap, even with async boundary (`import(...)`) between `index` and `bootstrap` and using remote only inside bootstrap, and the result was the same `initExternal('app')` was called as part of bundle's entry evaluation.

**Reply by zamotany** [reply]
I'm going to commit an example right now, you you'll be able to see the code itself.

**Reply by zamotany** [reply]
@ScriptedAlchemy @tom-sherman 
Here's the MF example I'm using:
- https://github.com/callstack/repack-examples/tree/main/module-federation - base
- https://github.com/callstack/repack-examples/tree/host-remotes/module-federation - with remotes in `host` package, this one fails on eager error

Once you clone repo, run `yarn install` and then `yarn workspaces module-federation run start`.

**Comment by tom-sherman** [comment]
IMO the ideal is that RePack would support `name@url` syntax out of the box without the `createRemote` call.

The reason I think this is ideal is that I believe module federation in react native should work as identically as possible to module federation on the web.

On the web ModuleFederationPlugin must transform the `@` syntax into a HTTP request, can't we inject a `loadScript` call in there?

**Reply by zamotany** [reply]
This will need some work in customising Webpack, because under the hood it will have to be a `promise new Promise` one way or another.

**Reply by tom-sherman** [reply]
I think that's the context I'm missing, could you point me to the code that transforms the @ syntax to pnp? This transform has to be different on the web vs node for example, so I don't see any technical reason why another target of react-native can't be added there.

**Reply by tom-sherman** [reply]
I can't get my head around how web, node, and Next can all work seamlessly but React Native can't.

**Reply by ScriptedAlchemy** [reply]
Im happy to provide any support for "native option parsing" 
Ive got like 7 plugins which do this so i can whip something up if you need any hands. 

All mine are usually

```js
apply(compiler) {
new ModuleFederationPlugin(convertedOptions).apply(compiler)
}
```

**Reply by zamotany** [reply]
I would happily point you to that code, if I knew where it is in Webpack.

**Reply by tom-sherman** [reply]
@ScriptedAlchemy so next and node.js achieve native @ syntax by extending MFP?

If so I think that's the way to go here too.

**Reply by ScriptedAlchemy** [reply]
Well remember, node, next.js are my custom implementations that use promise new promise to make it possible. Under the hood they do something similar to repack where i replace runtime targets with my own & have a augmented plugin that takes standard options and convert to promise new promise without user knowing.

**Reply by ScriptedAlchemy** [reply]
Yeah its native, as in the users just does what they would do in a browser, i then parse the options and apply the right thing depending on env. The idea is to provide a "just works" feel where all a user needs to do is maybe import another plugin like `NextJsModuleFederation(normalMFPoptions)` but never have to "learn" anything else beyond use this plugin and do what youd do anywhere else from a config standpoint

**Reply by tom-sherman** [reply]
Gotcha, I vote for the creation of `ReactNativeModuleFederationPlugin` then. As I understand it this would be some relatively simple string parsing that calls `createRemote` under the hood.

It's unclear to me how best to distribute this, should MFP be a peer dependency of RNMFP? Or simple a hard dependency?

**Reply by zamotany** [reply]
Creating a React Native specific plugin for Module Federation makes sense. I would put it inside Re.Pack under name `ReactNativeModuleFederationPlugin`, `RepackModuleFederationPlugin` or simply `ModuleFederationPlugin` so usage in code would look something like `Repack.ModuleFederationPlugin`. 

Publishing it separately from Re.Pack is not a good idea, because there's a lot more going on in Re.Pack just to make the Webpack-generated bundle even work in React Native.

> It's unclear to me how best to distribute this, should MFP be a peer dependency of RNMFP? Or simple a hard dependency?

`ModuleFederationPlugin` is a part of `webpack`, which is a peer dependency of Re.Pack, so there's noting more needed from end user once they've set up Re.Pack inside React Native project.

**Reply by ScriptedAlchemy** [reply]
Do you guys wanna go a google meet this week? Or a zoom? 

If we can hammer this out quickly, get a reproducible error repo, and talk shop. Then we can go out seperate ways and see what we come up with. Then add it back to this thread:)

**Reply by zamotany** [reply]
Yeah, I'm up for it.

**Reply by tom-sherman** [reply]
Same, any time after 11am UTC+1 is good for me. Zack just message me on slack with the details 👍

**Reply by zamotany** [reply]
You can find me on Discord or message me on https://twitter.com/_zamotany

**Reply by ScriptedAlchemy** [reply]
Great will get a twitter group convo going
